statemachine/signature.py

```diff
@@ -51,9 +52,12 @@ def wrap(cls, method):
 
         metadata_to_copy = method.func if isinstance(method, partial) else method
 
-        def method_wrapper(*args: Any, **kwargs: Any) -> Any:
+        async def method_wrapper(*args: Any, **kwargs: Any) -> Any:
             ba = sig_bind_expected(*args, **kwargs)
-            return method(*ba.args, **ba.kwargs)
+            if iscoroutinefunction(method):
+                return await method(*ba.args, **ba.kwargs)
+            else:
+                return method(*ba.args, **ba.kwargs)
```

Comment:
As this can contain blocking IO, I think that we may need to delegate to a thread worker.

Reply:
Would this work?
```suggestion
class SignatureAdapter(Signature):
    @classmethod
    def wrap(cls, method):
        """Build a wrapper that adapts the received arguments to the inner ``method`` signature"""

        sig = cls.from_callable(method)
        sig_bind_expected = sig.bind_expected

        metadata_to_copy = method.func if isinstance(method, partial) else method

        if iscoroutinefunction(method):

            async def method_wrapper(*args: Any, **kwargs: Any) -> Any:
                ba = sig_bind_expected(*args, **kwargs)
                return await method(*ba.args, **ba.kwargs)
        else:
            async def method_wrapper(*args: Any, **kwargs: Any) -> Any:
                ba = sig_bind_expected(*args, **kwargs)
                loop = asyncio.get_event_loop()
                with ThreadPoolExecutor() as pool:
                    return await loop.run_in_executor(pool, method, *ba.args, **ba.kwargs)

        method_wrapper.__name__ = metadata_to_copy.__name__

        return method_wrapper

```

Reply:
@rafaelrds measured a performance penalty of 3x slower execution when introducing the thread pool executor.

After considering this, we believe the current implementation should work well since the surrounding code and callbacks are written by the user.

If the surrounding code is asynchronous and the callback performs any blocking operations, the callback should ideally also be asynchronous and is a well know requirement for any Python developer writing async code. Conversely, if the StateMachine is working in a synchronous code base, then using synchronous callbacks is the expected behavior.